lib/spack/spack/directives.py

```diff
@@ -903,6 +917,29 @@ def _execute_requires(pkg):
     return _execute_requires
 
 
+@directive("vendors")
+def vendors(spec, when=None):
+    """Declares that a package has an internal copy of another package.
+
+    Currently, the effect is to forbid having the two packages in the same
+    "unification set".
+
+    Args:
+        spec: spec being vendored
+        when: optional constraint that triggers vendoring
+    """
```

Comment:
The only non-straightforward case for this would be one where we have properties on the conflicting package:
```python
conflicts("foo@3")
```
cause in that case we have to move properties checking to that package. Let me check how easy that is.

Reply:
Done in 7453f4c915083352ea7ce73f7da03877a7c8a299